sdk/python/kfp/dsl/_pipeline.py

```diff
@@ -108,7 +105,7 @@ def add_op(self, op: _container_op.ContainerOp, define_only: bool):
       op: An operator of ContainerOp or its inherited type.
     """
 
-    kubernetes_name = _make_kubernetes_name(op.human_name)
+    kubernetes_name = _sanitize_k8s_name(op.human_name)
```

Comment:
I missed the previous change, so I am adding my thoughts here: one design goal is to hide k8s as much as possible in "dsl" layer, and push the k8s stuff to compiler (I used to call it "argo compiler). This way the DSL layer is more generic, and that's why there are "dsl" and "compiler" separate directories. 

I feel like we don't have to sanitize the pipeline name here; We can store it as it is (so respect user's choice) and in compiler https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/compiler/compiler.py#L457 we can sanitize there. That way, we can move the util to compiler since it is very k8s specific.

WDYT?

Reply:
I'm all for the DSL hiding k8s. However, the pipeline sanitizes the name for operators which will be part of the final argo yaml. (in Pipeline.add_op() function).
I can sanitize the op names all in the compiler codes, though.

Reply:
Ah yes. Sorry I mixed pipeline name with step name. I think we can sanitize the name in compiler too as you mentioned.

That way, a different "compiler" may choose to sanitize it in a different way, or even not sanitize it at all.

Reply:
Done

Reply:
The `Pipeline` class (do not confuse with the `@pipeline` decorator) is a compiler helper class. It's only used during the compilation and it makes sense to sanitize the names/ids at this point. The original name remains untouched in `op.human_name`

Reply:
The Pipeline class is used in the @pipeline decorator and it might not be a good idea to move the Pipeline class to the Compiler because the dsl would depend on the compiler otherwise. Then there would be a dependency loop since the compiler depends on the dsl library. Right?

Reply:
> The Pipeline class is used in the @pipeline decorator

This is an implementation detail. We've talked with @qimingj and AFAIK he agreed that it was a good idea to unlink them like we did with `@python_component`. I've prepared the code for the most common case, but there was a slight problem for multi-pipeline file compilation. If we deprecate that feature (it's not currently used anywhere) we can unlink them easier.

Reply:
> That way, a different "compiler" may choose to sanitize it in a different way, or even not sanitize it at all.

I agree with that. That's why I moved the sanitization code from the `ContainerOp` to the compiler-relater `Pipeline` class. As you remember, I tried to detangle the `ContainerOp` from the `Pipeline` even more, by making it not required.

Reply:
Here is a proposal: Let's break the dependency between the DSL  classes and the compiler by adding generic events/hooks for events like `ContainerOp` creation or `@pipeline` application. The compiler can set the handlers for those hooks to execute some compiler-specific code. This way, the DSL does not depend on the compiler.